File: Arrays_and_Hashing/Python/49_group_anagrams.py
# RUNTIME: 92 ms, faster than 85.22% of python3 Submissions. 
# MEMORY USAGE: 19.6 MB, less than 76.26% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Create a dictionary
# 2. Iterate through the list
# 3. Sort the current element
# 4. Check if the sorted element is in the dictionary
# 5. If yes, append the current element to the dictionary
# 6. If no, create a new key with the sorted element and append the current element to the dictionary
# 7. Return the values of the dictionary

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        dict = {}
        ls = []
        for s in strs:
            sortedString = "".join(sorted(s))
            dict[sortedString] = dict.get(sortedString, []) +[s]
        for key in dict:
            ls.append(dict[key])
        return ls

File: Arrays_and_Hashing/C++/49_group_anagrams.cpp
/*
* RUNTIME: 39 ms - Beats 23.18% of submissions
* MEMORY - 21.3 MB - Beats 28.90% of submissions
*/

#include <vector>
#include <string>
#include <map>
#include <algorithm>
using namespace std;


/* Thoughtprocess:
*  1. Create a map of sorted strings to a vector of strings
*  2. Iterate through the input vector and sort each string
*  3. Add the sorted string to the map and push the original string to the vector
*  4. Iterate through the map and push the vectors to the result vector
* second indicates the element of the strings vector
*/


class Solution {
    public:
        vector<vector<string>> groupAnagrams(vector<string>& strs) {
            map<string, vector<string>> anagrams;
            vector<vector<string>> result;
            for (string s : strs)
            {
                string sorted = s;
                sort(sorted.begin(), sorted.end());
                anagrams[sorted].push_back(s);
            }
            for (auto idx : anagrams)
            {
                result.push_back(idx.second);
            }
            return result;
        }
};

